src/main/java/de/tum/in/www1/artemis/web/rest/competency/CourseCompetencyResource.java

```diff
@@ -2,45 +2,160 @@
+        this.irisCompetencyGenerationSessionService = irisCompetencyGenerationSessionService;
+        this.competencyJolService = competencyJolService;
+        this.authorizationCheckService = authorizationCheckService;
+    }
+
+    /**
+     * GET /course-competencies/:competencyId/title : Returns the title of the course competency with the given id
+     *
+     * @param competencyId the id of the course competency
+     * @return the title of the course competency wrapped in an ResponseEntity or 404 Not Found if no competency with that id exists
+     */
+    @GetMapping("course-competencies/{competencyId}/title")
+    @EnforceAtLeastStudent
+    public ResponseEntity<String> getCompetencyTitle(@PathVariable long competencyId) {
+        final var title = courseCompetencyRepository.getCompetencyTitle(competencyId);
+        return title == null ? ResponseEntity.notFound().build() : ResponseEntity.ok(title);
     }
 
     /**
-     * GET courses/:courseId/competencies : gets all the course competencies of a course
+     * GET courses/{courseId}/course-competencies/titles : Returns the titles of all course competencies. Used for a validator in the client
+     *
+     * @param courseId the id of the current course
+     * @return the titles of all course competencies
+     */
+    @GetMapping("courses/{courseId}/course-competencies/titles")
+    @EnforceAtLeastEditorInCourse
+    public ResponseEntity<List<String>> getCourseCompetencyTitles(@PathVariable Long courseId) {
+        final var titles = courseCompetencyRepository.findAllTitlesByCourseId(courseId);
+        return ResponseEntity.ok(titles);
+    }
+
+    /**
+     * GET courses/:courseId/competencies/:competencyId : gets the competency with the specified id including its related exercises and lecture units
+     * This method also calculates the user progress
+     *
+     * @param competencyId the id of the competency to retrieve
+     * @param courseId     the id of the course to which the competency belongs
+     * @return the ResponseEntity with status 200 (OK) and with body the competency, or with status 404 (Not Found)
+     */
+    @GetMapping("courses/{courseId}/course-competencies/{competencyId}")
+    @EnforceAtLeastStudentInCourse
+    public ResponseEntity<CourseCompetency> getCourseCompetency(@PathVariable long competencyId, @PathVariable long courseId) {
+        log.info("REST request to get Competency : {}", competencyId);
+        var currentUser = userRepository.getUserWithGroupsAndAuthorities();
+        var course = courseRepository.findByIdElseThrow(courseId);
+        var competency = courseCompetencyService.findCompetencyWithExercisesAndLectureUnitsAndProgressForUser(competencyId, currentUser.getId());
+        checkCourseForCompetency(course, competency);
+
+        courseCompetencyService.filterOutLearningObjectsThatUserShouldNotSee(competency, currentUser);
+
+        return ResponseEntity.ok().body(competency);
```

Comment:
(Simplicity) Could we instead write 

```suggestion
        return ResponseEntity.ok(competency);
```